### PR TITLE
Fix ReferenceHandler test failures when enabling small buffer async in tests

### LINKED ISSUE #66727: System.Text.Json ReferenceHandler test failures when enabling async in tests
Opening issue to track test failures encountered when enabling async stress testing in our ReferenceHandler unit tests as part of the work for #66463. Here are a few examples:

```
Failed System.Text.Json.Serialization.Tests.ReferenceHandlerTestsDynamic_AsyncStream.JsonPathIncompletePropertyAfterMetadata [17 ms]
  Error Message:
   Assert.Throws() Failure
Expected: typeof(System.Text.Json.JsonException)
Actual:   typeof(System.NullReferenceException): Object reference not set to an instance of an object.
---- System.NullReferenceException : Object reference not set to an instance of an object.
  Stack Trace:
     at System.Text.Json.ReadStack.JsonPath() in C:\Users\eitsarpa\devel\dotnet\clonedruntime2\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\ReadStack.cs:line 216
   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, JsonReaderException ex) in C:\Users\eitsarpa\devel\dotnet\clonedruntime2\src\libraries\System.Text.Json\src\System\Text\Json\ThrowHelper.Serialization.cs:line 246
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state) in C:\Users\eitsarpa\devel\dotnet\clonedruntime2\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\JsonConverterOfT.ReadCore.cs:line 79
```

```
Failed System.Text.Json.Serialization.Tests.ReferenceHandlerTestsDynamic_AsyncStream.PreservedArrayValuesContainsNull [< 1 ms]
  Error Message:
   Assert.Equal() Failure
           ↓ (pos 1)
Expected: $.$values
Actual:   $
           ↑ (pos 1)
  Stack Trace:
     at System.Text.Json.Serialization.Tests.ReferenceHandlerTests.PreservedArrayValuesContainsNull() in C:\Users\eitsarpa\devel\dotnet\clonedruntime2\src\libraries\System.Text.Json\tests\Common\ReferenceHandlerTests\ReferenceHandlerTests.Deserialize.cs:line 1406
```
These are likely product bugs that should be investigated.

Related to work for https://github.com/dotnet/runtime/issues/63747

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Enables small buffer async in the ReferenceHandler test suite. Fixes a couple of bugs that were causing tests to fail:

* Fix `JsonPath()` methods throwing NullReferenceExceptions when `_count == 0 && _continuationCount == 1`.
* Temporarily address a corner case in the metadata reader async state machine implementation. To be addressed properly in an upcoming refactoring of the metadata reader components.

Fix #66727.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs [hunk 1]
```diff
@@ -336,6 +336,11 @@ public static partial class JsonSerializer
 
             if (state.Current.ObjectState == StackFrameObjectState.ReadValuesStartArray)
             {
+                // Temporary workaround for the state machine accidentally
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:339]: Temporary until what?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:339]: > Temporarily address a corner case in the metadata reader async state machine implementation. To be addressed properly in an upcoming refactoring of the metadata reader components.

I'm rewriting this particular code in an upcoming PR, but needed to get the new test coverage green beforehand.
